brdnet/subset_studies.py

```diff
@@ -0,0 +1,284 @@
+
+    data = requests.get(query_string).json()
+
+    # Collect all term ids associated with the current study
+    current_study_terms = data['terms']
+    term_set = set()
+    for term in current_study_terms:
+        ids = term['dterm']['ids']
+        for term_id in ids:
+            term_set.add(term_id)
+
+    for term in required_terms:
+        if term not in term_set:
+            return False
+    return True
+
+
+def subset_df(df, terms):
+    '''Use MetaSRA to filter the dataframe so that it only contains studies matching all the
+    terms passed in as arguments
+
+    Arguments
+    ---------
+    df: pandas.DataFrame
+        The dataframe to be subsetted
+    terms: list of strs
+        The ontology ids to be used in filtering the dataframe. For example: 'UBERON:0000955' would
+        be passed in if the user only wanted samples from brains
+
+    Returns
+    -------
+    subset_df: pandas.DataFrame
+        The subset of df that matches all the terms
+    '''
+    samples = df.columns
+    # Get the study id for each sample
+    studies = [sample.strip().split('.')[0] for sample in samples]
+    # Keep only one instance of each study id
+    studies = list(set(studies))
+
+    studies_to_keep = []
+    for study in studies:
+        if study_has_all_terms(study, terms):
+            studies_to_keep.append(study)
+
+    # Keep only samples from studies in studies_to_keep
+    filtered_studies = get_df_subset_by_study(df, studies_to_keep)
+
+    return filtered_studies
+
+
+def get_study_counts(sample_list):
+    '''Get the names of all studies in the list and the number of samples each one has
+
+    Arguments
+    ---------
+    sample_list: list of strs
+        A list of samples in the 'SRPxxxxx.SRRxxxxxxx' format
+
+    Returns
+    -------
+    Counter
+        A Counter object mapping studies to sample counts
+    '''
+    studies = [sample.strip().split('.')[0] for sample in sample_list]
+    return collections.Counter(studies)
+
+
+def balance_study_counters(plier_study_counter, classifier_study_counter, fraction):
+    done_balancing = False
+
+    plier_sample_count = sum(plier_study_counter.values())
+    classifier_sample_count = sum(classifier_study_counter.values())
+
+    total_samples = plier_sample_count + classifier_sample_count
+    plier_target = int(total_samples * fraction)
+    classifier_target = total_samples - plier_target
+
+    # Go back and forth moving studies from one set to the other if it decreases the total error
+    while not done_balancing:
+        old_plier_count = plier_sample_count
+
+        error = abs(plier_target - plier_sample_count)
+
+        # We have to cast these as lists because python hates it when you modify a dictionary
+        # as you are iterating over it
+        plier_study_list = list(plier_study_counter.keys())
+        for study in plier_study_list:
+            # If we already have too few samples from plier, don't look to move any over
```

Comment:
Is there any way you can abstract some of this logic into a function? It's a bit hard for me to look at this code and know exactly what's going on in these inner loops, although I understand the overall goal of `balance_study_counters`.

My feeling is that moving some of this code into a function (even if it's only called once) would help to clearly delineate what the inputs and outputs are here, which I think would make `balance_study_counters` as a whole easier to read and understand.

Feel free to make a separate issue for this if you think it'll take a lot of effort, since it seems like things are working fine now.

Reply:
Good call, I split up the logic into the functions `update_study_counters`, `move_plier_to_classifier`, and `move_classifier_to_plier`, so it should be much more readable now.